Raw pointer operations

DemiMarie opened the pull request:
Both Haskell and MLton offer operations on raw pointers.  These
operations are unsafe, but libraries like Haskell's Bytestring wrap them
in a safe and efficient interface.  Furthermore, Jane Street maintained
raw pointer support in their internal compiler fork, and it is already
possible to implement raw pointer operations by either writing C
primitives or abusing existing unsafe primitives.  Finally, raw pointers
allow inspecting and manipulating C data structures directly from OCaml,
which is faster than using "alloc" C functions to allocate OCaml data
structures from C.

This PR adds compiler, runtime, and standard library support for raw
pointer operations.  To avoid needing a bootstrap, the primitives are
normal primitives, not %-primitives, so that the bootstrap compiler will
accept them and (correctly) compile them to C calls.  The compilers know
about these primitives, and the native code compiler generates efficient
machine code without calling any C functions.  For this reason, the
primitives are in a new file (pointer.c) that is only built for bytecode.

DemiMarie committed:
Raw pointer operations (requires bootstrap)

Both Haskell and MLton offer operations on raw pointers.  These
operations are unsafe, but libraries like Haskell's Bytestring wrap them
in a safe and efficient interface.  Furthermore, Jane Street maintained
raw pointer support in their internal compiler fork, and it is already
possible to implement raw pointer operations by either writing C
primitives or abusing existing unsafe primitives.  Finally, raw pointers
allow inspecting and manipulating C data structures directly from OCaml,
which is faster than using "alloc" C functions to allocate OCaml data
structures from C.

This commit adds compiler, runtime, and standard library support for raw
pointer operations.  The compilers know about these primitives, and the
native code compiler generates efficient machine code without calling
any C functions.  For this reason, the primitives are in a new file
(pointer.c) that is only built for bytecode.

This requires a bootstrap due to the new %-primitives.

nojb commented:
This PR (or something very close to it) has already beed proposed in the past several times and rejected (#652, #7252, #724, #9910), mostly due to a strong objection to the OCaml core system exposing this kind of low-level operations: https://github.com/ocaml/ocaml/issues/7252#issuecomment-473059947.

Has something changed? Otherwise, I do not see how we could expect a different outcome...

Octachron commented:
Since this an already rejected feature wish without any further arguments to support it nor references to previous discussions (and with even more possibility of segfaults than previous propositions), I propose to close this PR right now to avoid losing time.

lthls commented:
You can already implement something almost as efficient with the existing primitives (only one additional subtraction):
```ocaml
(* A statically allocated string will always have the same address *)
let base = ""
let base_address = Obj.raw_field (Obj.repr (Some base)) 0

type t = nativeint

(* Export the unsafe string access primitives
   Note: the setters would also be exported with type string, not bytes *)
external string_get64u : string -> int -> int64 = "%caml_string_get64u"

(* Implement raw loads in terms of string accesses.
   Offsets are in bytes. *)
let load64 ptr ofs =
  string_get64u base (ofs + Nativeint.(to_int (sub ptr base_address)))

(* Test *)
let x = [| 1; 2; 3; 4 |]
let ptr = Obj.raw_field (Obj.repr (Some x)) 0
let () =
  Format.printf "[| %Ld; %Ld; %Ld; %Ld |]@."
    (load64 ptr 0)
    (load64 ptr 8)
    (load64 ptr 16)
    (load64 ptr 24)
```

DemiMarie commented:
> You can already implement something almost as efficient with the existing primitives (only one additional subtraction):
> 
> ```ocaml
> (* A statically allocated string will always have the same address *)
> let base = ""
> let base_address = Obj.raw_field (Obj.repr (Some base)) 0
> 
> type t = nativeint
> 
> (* Export the unsafe string access primitives
>    Note: the setters would also be exported with type string, not bytes *)
> external string_get64u : string -> int -> int64 = "%caml_string_get64u"
> 
> (* Implement raw loads in terms of string accesses.
>    Offsets are in bytes. *)
> let load64 ptr ofs =
>   string_get64u base (ofs + Nativeint.(to_int (sub ptr base_address)))
> 
> (* Test *)
> let x = [| 1; 2; 3; 4 |]
> let ptr = Obj.raw_field (Obj.repr (Some x)) 0
> let () =
>   Format.printf "[| %Ld; %Ld; %Ld; %Ld |]@."
>     (load64 ptr 0)
>     (load64 ptr 8)
>     (load64 ptr 16)
>     (load64 ptr 24)
> ```

Is this guaranteed to work in future versions of the compiler?

> Since this an already rejected feature wish without any further arguments to support it nor references to previous discussions (and with even more possibility of segfaults than previous propositions), I propose to close this PR right now to avoid losing time.

Back in 2020 Jane Street needed something like this for their own internal use (see #9910).  They wound up maintaining a downstream compiler patch, which isn’t great at all.  I’d much rather them _not_ need that.

This could also allow implementing the Bigarray primitives as library code, replacing the existing %-primitives used there.

lthls commented:
> Is this guaranteed to work in future versions of the compiler?

As much as any other `Obj`-using code, I would say. Although the test that I've provided is not safe, as `x` could be heap-allocated and have its address move between the definition of `ptr` and its uses; a proper test would use a bigarray but I was lazy.

DemiMarie committed:
Optimize 32-bit stores

These do not care about the top 32 bits, so sign and zero extension
operations can be omitted.

DemiMarie commented:
The bootstrap can be eliminated by replacing the `%` primitives with normal primitives that happen to be in the primitive table.  The boostrap and post-bootstrap `ocamlc` will both compile them as C calls, while the post-bootstrap `ocamlopt` will produce native code for them.

DemiMarie commented:
> The bootstrap can be eliminated by replacing the `%` primitives with normal primitives that happen to be in the primitive table. The boostrap and post-bootstrap `ocamlc` will both compile them as C calls, while the post-bootstrap `ocamlopt` will produce native code for them.

Turns out this trick doesn’t work.

DemiMarie committed:
Bootstrap

DemiMarie commented:
> You can already implement something almost as efficient with the existing primitives (only one additional subtraction):
> 
> ```ocaml
> (* A statically allocated string will always have the same address *)
> let base = ""
> let base_address = Obj.raw_field (Obj.repr (Some base)) 0
> 
> type t = nativeint
> 
> (* Export the unsafe string access primitives
>    Note: the setters would also be exported with type string, not bytes *)
> external string_get64u : string -> int -> int64 = "%caml_string_get64u"
> 
> (* Implement raw loads in terms of string accesses.
>    Offsets are in bytes. *)
> let load64 ptr ofs =
>   string_get64u base (ofs + Nativeint.(to_int (sub ptr base_address)))
> 
> (* Test *)
> let x = [| 1; 2; 3; 4 |]
> let ptr = Obj.raw_field (Obj.repr (Some x)) 0
> let () =
>   Format.printf "[| %Ld; %Ld; %Ld; %Ld |]@."
>     (load64 ptr 0)
>     (load64 ptr 8)
>     (load64 ptr 16)
>     (load64 ptr 24)
> ```

I think this won’t work if `sub ptr base_address` wraps around.

DemiMarie commented:
@xavierleroy: Is this just an expanded version of #9655?  #9655 already allows one to peek and poke already values of type `uintptr_t`, and this (effectively) just expands that to other types.

If you’re still opposed, feel free to close this.
